packages/core/src/lifecycle-registry.ts

```diff
@@ -0,0 +1,211 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Binding, BindingScope, BindingType, Context} from '@loopback/context';
+import {CoreBindings} from './keys';
+import {LifeCycleObserver} from './lifecycle';
+const CoreTags = CoreBindings.Tags;
+import debugFactory = require('debug');
+const debug = debugFactory('loopback:core:lifecycle');
+
+/**
+ * A group of life cycle observers
+ */
+export type LifeCycleObserverGroup = {
+  group: string;
+  bindings: Readonly<Binding<LifeCycleObserver>>[];
+};
+
+export type LifeCycleObserverOptions = {
+  /**
+   * Notify observers of the same group in parallel, default to `true`
+   */
+  parallel?: boolean;
+};
+
+/**
+ *
+ * A Context subclass that supports life cycle
+ *
+ */
+export class LifeCycleObserverRegistry implements LifeCycleObserver {
+  constructor(
+    protected ctx: Context,
+    protected options: LifeCycleObserverOptions = {parallel: true},
+  ) {}
+  /**
+   * Configures the ordering of life cycle observers by group names
+   */
+  protected groupsByOrder: string[] = ['server'];
+
+  setGroupsByOrder(groups: string[]) {
+    this.groupsByOrder = groups || ['server'];
+  }
+
+  /**
+   * Find all life cycle observer bindings. By default, a constant or singleton
+   * binding tagged with `CoreBindings.Tags.LIFE_CYCLE_OBSERVER` or
+   * `CoreBindings.Tags.SERVER`.
+   */
+  findObserverBindings() {
+    return this.ctx.find<LifeCycleObserver>(
+      binding =>
+        (binding.type === BindingType.CONSTANT ||
+          binding.scope === BindingScope.SINGLETON) &&
+        (binding.tagMap[CoreTags.LIFE_CYCLE_OBSERVER] != null ||
+          binding.tagMap[CoreTags.SERVER]),
+    );
+  }
+
+  /**
+   * Get observer groups by order
+   */
+  getObserverGroupsByOrder(): LifeCycleObserverGroup[] {
+    const bindings = this.findObserverBindings();
+    return this.sortObserverBindingsByGroup(bindings);
+  }
+
+  /**
+   * Get the group for a given life cycle observer binding
+   * @param binding Life cycle observer binding
+   */
+  protected getObserverGroup(
+    binding: Readonly<Binding<LifeCycleObserver>>,
+  ): string {
+    // First check if there is an explicit group name in the tag
+    let group = binding.tagMap[CoreTags.LIFE_CYCLE_OBSERVER_GROUP];
+    if (!group) {
+      // Fall back to a tag that matches one of the groups
+      group = this.groupsByOrder.find(g => binding.tagMap[g] === g);
+    }
+    return group || '';
+  }
+
+  /**
+   * Sort the life cycle observer bindings so that we can start/stop them
+   * in the right order. By default, we can start other observers before servers
```

Comment:
I don't understand why is this sorting still required when we already have `preStart`/`start`/`postStart` groups allowing observers to decide whether they want to run before or after the server is started. Could you please explain? What am I missing?